File: 백준/Gold/2473. 세 용액/세 용액.js
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];
let N, sequence;

readline.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  N = Number(input[0]);
  sequence = input[1].split(' ').map(Number);
  const result = solution();
  console.log(result.join(' '));
  process.exit();
});

const solution = () => {
  let result = [Infinity, Infinity, Infinity];
  let minSum = Infinity;

  sequence.sort((a, b) => a - b);
  
  for (let i = 0; i < N; i++) {
    let left = 0, right = N-1;
    while (left < right) {
      left = (i === left ? left+1 : left);
      right = (i === right ? right-1 : right);

      if (left === right) continue;

      const sum = sequence[i]+sequence[left]+sequence[right];
      if (sum > 0) {
        if (minSum > Math.abs(sum)) {
          minSum = Math.abs(sum);
          result = [sequence[i], sequence[left], sequence[right]];
        }
        right--;
      } else if (sum < 0) {
        if (minSum > Math.abs(sum)) {
          minSum = Math.abs(sum);
          result = [sequence[i], sequence[left], sequence[right]];
        }
        left++;
      } else {
        return [sequence[i], sequence[left], sequence[right]].sort((a, b) => a - b);
      }
    }
  }

  return result.sort((a, b) => a - b);
}

File: 백준/Gold/2473. 세 용액/세 용액.py
import sys 
input = sys.stdin.readline

N = int(input())
array = list(map(int, input().split())) 

array.sort() 
result = [array[-3], array[-2], array[-1]]

for index, point in enumerate(array):
  left = 0 if index != 0 else 1
  right = N-1 if index != N-1 else N-2
  while left < right and right < N:
    current = array[left]+array[right]+point  
    if abs(current-0) < abs(sum(result)-0):
      result = [array[left], array[right], point]
    if current > 0:
      right -= 1
      if index == right: right -= 1
    elif current < 0:
      left += 1
      if index == left: left += 1
    else:
      break

result.sort()
print(' '.join(map(str, result)))